packages/xpetra/sup/Utils/Xpetra_MatrixMatrix.hpp

```diff
@@ -561,68 +561,19 @@ Note: this class is not in the Xpetra_UseShortNames.hpp
-
-            } else {
-              for (size_t i = 0; i < numLocalRows; ++i)
-                exactNnzPerRow[i] = A.getNumEntriesInLocalRow(Teuchos::as<LO>(i)) + maxNzInB;
-            }
-
-            fos << "MatrixMatrix::TwoMatrixAdd : special case detected (one matrix has a fixed nnz per row)"
-                << ", using static profiling" << std::endl;
-            C = rcp(new CrsMatrixWrap(A.getRowMap(), exactNnzPerRow, Xpetra::StaticProfile));
-          } else {
-            // general case
-            double nnzPerRowInA = Teuchos::as<double>(A.getNodeNumEntries()) / A.getNodeNumRows();
-            double nnzPerRowInB = Teuchos::as<double>(B.getNodeNumEntries()) / B.getNodeNumRows();
-            LO    nnzToAllocate = Teuchos::as<LO>( (nnzPerRowInA + nnzPerRowInB) * 1.5) + Teuchos::as<LO>(1);
-
-            LO maxPossible = A.getNodeMaxNumRowEntries() + B.getNodeMaxNumRowEntries();
-            //Use static profiling (more efficient) if the estimate is at least as big as the max
-            //possible nnz's in any single row of the result.
-            Xpetra::ProfileType pft = (maxPossible) > nnzToAllocate ? Xpetra::DynamicProfile : Xpetra::StaticProfile;
-
-            fos << "nnzPerRowInA = " << nnzPerRowInA << ", nnzPerRowInB = " << nnzPerRowInB << std::endl;
-            fos << "MatrixMatrix::TwoMatrixAdd : space allocated per row = " << nnzToAllocate
-                << ", max possible nnz per row in sum = " << maxPossible
-                << ", using " << (pft == Xpetra::DynamicProfile ? "dynamic" : "static" ) << " profiling"
-                << std::endl;
-            C = rcp(new CrsMatrixWrap(A.getRowMap(), nnzToAllocate, pft));
-          }
-          if (transposeB)
-            fos << "MatrixMatrix::TwoMatrixAdd : ** WARNING ** estimate could be badly wrong because second summand is transposed" << std::endl;
-        }
-
         if (C->getRowMap()->lib() == Xpetra::UseEpetra) {
           throw Exceptions::RuntimeError("MatrixMatrix::Add for Epetra only available with Scalar = double, LO = GO = int.");
         } else if (C->getRowMap()->lib() == Xpetra::UseTpetra) {
   #ifdef HAVE_XPETRA_TPETRA
           const Tpetra::CrsMatrix<SC,LO,GO,NO>& tpA = Xpetra::Helpers<SC,LO,GO,NO>::Op2TpetraCrs(A);
           const Tpetra::CrsMatrix<SC,LO,GO,NO>& tpB = Xpetra::Helpers<SC,LO,GO,NO>::Op2TpetraCrs(B);
           RCP<Tpetra::CrsMatrix<SC,LO,GO,NO> >  tpC = Xpetra::Helpers<SC,LO,GO,NO>::Op2NonConstTpetraCrs(C);
-
-          Tpetra::MatrixMatrix::Add(tpA, transposeA, alpha, tpB, transposeB, beta, tpC);
+          auto addParams = rcp(new Teuchos::ParameterList);
+          addParams->set<bool>("Call fillComplete", false);
+          //passing A after B means C will have the same domain/range map as A (or A^T if transposeA)
+          Tpetra::MatrixMatrix::add<SC,LO,GO,NO>(beta, transposeB, tpB, alpha, transposeA, tpA, *tpC, Teuchos::null, Teuchos::null, addParams);
```

Comment:
There are two more specializations of `TwoMatrixAdd` that also need to be changed..

Reply:
@cgcgcg Those versions of Xpetra::TwoMatrixAdd() can take a non fill-complete B, and they add values into B (rather than making a new C). Tpetra::MatrixMatrix::add() doesn't support this yet.

Reply:
How about these guys?
https://github.com/trilinos/Trilinos/blob/75333e4ed9915f29cfcd916dc01ec877e81b60fc/packages/xpetra/sup/Utils/Xpetra_MatrixMatrix.hpp#L1254-L1256
https://github.com/trilinos/Trilinos/blob/75333e4ed9915f29cfcd916dc01ec877e81b60fc/packages/xpetra/sup/Utils/Xpetra_MatrixMatrix.hpp#L1870-L1872

Reply:
Yep you're right.